packages/gatsby-plugin-utils/src/polyfill-remote-file/utils/url-generator.ts

```diff
@@ -1,28 +1,75 @@
+import { basename, extname } from "path"
+import { URL } from "url"
+import { createContentDigest } from "gatsby-core-utils/create-content-digest"
 import { isImage } from "../types"
 import type { ImageCropFocus, WidthOrHeight } from "../types"
 
-export function generatePublicUrl(
-  {
-    url,
-    mimeType,
-  }: {
-    url: string
-    mimeType: string
-  },
-  checkMimeType: boolean = true
+const ORIGIN = `https://gatsbyjs.com`
+
+export enum ImageCDNUrlKeys {
+  URL = `u`,
+  ARGS = `a`,
+}
+
+export function generateFileUrl({
+  url,
+  filename,
+}: {
+  url: string
+  filename: string
+}): string {
+  const fileExt = extname(filename)
+  const filenameWithoutExt = basename(filename, fileExt)
+
+  const parsedURL = new URL(
+    `${ORIGIN}${generatePublicUrl({
+      url,
+    })}/${filenameWithoutExt}${fileExt}`
+  )
+
+  parsedURL.searchParams.append(ImageCDNUrlKeys.URL, url)
+
+  return `${parsedURL.pathname}${parsedURL.search}`
+}
+
+export function generateImageUrl(
+  source: { url: string; mimeType: string; filename: string },
+  imageArgs: Parameters<typeof generateImageArgs>[0]
 ): string {
-  const remoteUrl = Buffer.from(url).toString(`base64`)
+  const filenameWithoutExt = basename(source.filename, extname(source.filename))
+
+  const parsedURL = new URL(
+    `${ORIGIN}${generatePublicUrl(source)}/${createContentDigest(
+      generateImageArgs(imageArgs)
```

Comment:
@wardpeet Clever to use the md5 hash of the args (for ease of creating unique and short file paths) and also adding it as unhashed query params (for use in gatsby cloud router)!